pkg/registry/relocator.go

```diff
@@ -10,76 +10,114 @@ import (
 
 	"github.com/google/go-containerregistry/pkg/authn"
 	"github.com/google/go-containerregistry/pkg/name"
-	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1"
 	"github.com/google/go-containerregistry/pkg/v1/remote"
-	"github.com/pkg/errors"
 )
 
+type RelocateCfg struct {
+	imgInfo         ImageInfo
+	imgWriteOptions []remote.Option
+}
+
+type ImageInfo struct {
+	refRepo      name.Reference
+	refDigestStr string
+	tag          name.Tag
+	size         int64
+}
+
+type DryRunRelocator struct{}
+
+func (d DryRunRelocator) Relocate(srcImage v1.Image, dstRepoStr string, writer io.Writer, tlsCfg TLSConfig) (string, error) {
+	cfg, err := getRelocateCfg(srcImage, dstRepoStr, tlsCfg)
+	writer.Write([]byte(fmt.Sprintf("\tUploading '%s'\n", cfg.imgInfo.refDigestStr)))
+	return cfg.imgInfo.refDigestStr, err
+}
+
 type Relocator struct{}
 
-func (r *Relocator) Relocate(writer io.Writer, image v1.Image, dest string, tlsCfg TLSConfig) (string, error) {
-	ref, err := name.ParseReference(dest, name.WeakValidation)
+func (r Relocator) Relocate(srcImage v1.Image, dstRepoStr string, writer io.Writer, tlsCfg TLSConfig) (string, error) {
+	cfg, err := getRelocateCfg(srcImage, dstRepoStr, tlsCfg)
 	if err != nil {
-		return "", errors.WithStack(err)
+		return "", err
 	}
 
-	refName := fmt.Sprintf("%s/%s", ref.Context().RegistryStr(), ref.Context().RepositoryStr())
-	ref, err = name.ParseReference(refName, name.WeakValidation)
-	if err != nil {
-		return "", errors.WithStack(err)
-	}
+	i := cfg.imgInfo
+	writer.Write([]byte(fmt.Sprintf("\tUploading '%s'", i.refDigestStr)))
 
-	digest, err := image.Digest()
-	if err != nil {
-		return "", errors.WithStack(err)
-	}
+	spinner := NewUploadSpinner(writer, i.size)
+	defer spinner.Stop()
+	go spinner.Write()
 
-	size, err := imageSize(image)
+	err = remote.Write(i.refRepo, srcImage, cfg.imgWriteOptions...)
 	if err != nil {
-		return "", errors.WithStack(err)
+		err = NewImageAccessError(i.refRepo.Context().RegistryStr(), err)
+		return i.refDigestStr, err
```

Comment:
```suggestion
		return i.refDigestStr, NewImageAccessError(i.refRepo.Context().RegistryStr(), err)
```

Reply:
marking as resolved. addressed in commit https://github.com/vmware-tanzu/kpack-cli/commit/be211733ecaa33231703f7b9d2938b0b71bee3a3
please reopen if not ok